File: Api/IntegrationAzure.Api/Application/Services/LogService.cs
using IntegrationAzure.Api.Application.DTOs;
using IntegrationAzure.Api.Domain.Entities;
using IntegrationAzure.Api.Domain.Interfaces;

namespace IntegrationAzure.Api.Application.Services;

/// <summary>
/// Serviço para operações relacionadas a logs
/// Implementa regras de negócio e orchestração
/// </summary>
public class LogService
{
    private readonly ILogRepository _logRepository;

    public LogService(ILogRepository logRepository)
    {
        _logRepository = logRepository ?? throw new ArgumentNullException(nameof(logRepository));
    }

    /// <summary>
    /// Obtém logs com filtros
    /// </summary>
    public async Task<ApiResponseDto<List<LogSummaryDto>>> GetFilteredAsync(LogFilterDto filter)
    {
        try
        {
            var logs = await _logRepository.GetAllAsync();

            // Aplicar filtros
            if (!string.IsNullOrEmpty(filter.UserId))
                logs = logs.Where(l => l.UserId.Contains(filter.UserId));

            if (!string.IsNullOrEmpty(filter.Entity))
                logs = logs.Where(l => l.Entity.Contains(filter.Entity));

            if (filter.Level.HasValue)
                logs = logs.Where(l => l.Level == filter.Level);

            if (filter.StartDate.HasValue)
                logs = logs.Where(l => l.CreatedAt >= filter.StartDate);

            if (filter.EndDate.HasValue)
                logs = logs.Where(l => l.CreatedAt <= filter.EndDate);

            // Ordenar por data decrescente (mais recentes primeiro)
            logs = logs.OrderByDescending(l => l.CreatedAt);

            // Paginação
            var totalCount = logs.Count();
            var pagedLogs = logs
                .Skip((filter.PageNumber - 1) * filter.PageSize)
                .Take(filter.PageSize)
                .Select(MapToSummaryDto)
                .ToList();

            return new ApiResponseDto<List<LogSummaryDto>>
            {
                Success = true,
                Message = $"{totalCount} logs encontrados",
                Data = pagedLogs
            };
        }
        catch (Exception ex)
        {
            return new ApiResponseDto<List<LogSummaryDto>>
            {
                Success = false,
                Message = "Erro interno do servidor",
                Errors = new List<string> { ex.Message }
            };
        }
    }

    /// <summary>
    /// Método de conveniência para criar logs rapidamente
    /// </summary>
    public async Task LogActionAsync(string action, string entity, string? entityId, string userId, string? details = null, Domain.Entities.LogLevel level = Domain.Entities.LogLevel.Info)
    {
        try
        {
            var log = new Log
            {
                Action = action,
                Entity = entity,
                EntityId = entityId,
                UserId = userId,
                Details = details,
                Level = level,
                CreatedBy = userId
            };

            await _logRepository.AddAsync(log);
            await _logRepository.SaveChangesAsync();
        }
        catch
        {
            // Silenciosamente falha - logs não devem quebrar a aplicação
        }
    }

    /// <summary>
    /// Mapeia entidade para DTO de resumo
    /// </summary>
    private static LogSummaryDto MapToSummaryDto(Log log)
    {
        return new LogSummaryDto
        {
            Id = log.Id,
            Action = log.Action,
            Entity = log.Entity,
            EntityId = log.EntityId,
            UserId = log.UserId,
            Level = log.Level,
            CreatedAt = log.CreatedAt
        };
    }
}

File: Aplication/src/services/logService.js
import api from './api';

/**
 * Serviço para operações com logs
 * Integra com os endpoints da API .NET
 */
export const logService = {
    /**
     * Obtém logs com filtros
     */
    async getLogs(filters = {}) {
        try {
            const params = new URLSearchParams();

            if (filters.userId) params.append('userId', filters.userId);
            if (filters.entity) params.append('entity', filters.entity);
            if (filters.level) params.append('level', filters.level);
            if (filters.startDate) params.append('startDate', filters.startDate);
            if (filters.endDate) params.append('endDate', filters.endDate);
            if (filters.pageSize) params.append('pageSize', filters.pageSize);
            if (filters.pageNumber) params.append('pageNumber', filters.pageNumber);

            const response = await api.get(`/logs?${params.toString()}`);
            return response.data;
        } catch (error) {
            throw this.handleError(error);
        }
    },

    /**
     * Trata erros das requisições
     */
    handleError(error) {
        if (error.response) {
            // Erro de resposta da API
            const { data, status } = error.response;
            return {
                message: data.message || 'Erro na operação solicitada',
                errors: data.errors || ['Erro interno do sistema'],
                status
            };
        } else if (error.request) {
            // Erro de rede
            return {
                message: 'Erro de conexão com o servidor',
                errors: ['Verifique sua conexão com a internet e tente novamente'],
                status: 0
            };
        } else {
            // Outro tipo de erro
            return {
                message: 'Erro inesperado',
                errors: ['Ocorreu um erro inesperado. Tente novamente.'],
                status: 0
            };
        }
    }
};
